lib/protoparser/zabbixconnector/parser.go

```diff
@@ -0,0 +1,267 @@
+	rs.Rows = unmarshalRows(rs.Rows[:0], s, par)
+}
+
+// Row is a single Zabbix Connector row
+type Row struct {
+	Tags      []Tag
+	Value     float64
+	Timestamp int64
+}
+
+func (r *Row) reset() {
+	tags := r.Tags
+	for i := range tags {
+		tags[i].reset()
+	}
+	r.Tags = tags[:0]
+
+	r.Value = 0
+	r.Timestamp = 0
+}
+
+func (r *Row) addTag() *Tag {
+	dst := r.Tags
+	if cap(dst) > len(dst) {
+		dst = dst[:len(dst)+1]
+	} else {
+		dst = append(dst, Tag{})
+	}
+	tag := &dst[len(dst)-1]
+	r.Tags = dst
+	return tag
+}
+
+func (r *Row) unmarshal(o *fastjson.Value, par byte) error {
+	r.reset()
+
+	host := o.GetObject("host")
+	if host == nil {
+		return fmt.Errorf("missing `host` object")
+	}
+
+	v := host.Get("host").GetStringBytes()
+	if len(v) == 0 {
+		return fmt.Errorf("missing `host` element in `host` object")
+	}
+	tag := r.addTag()
+	tag.Key = append(tag.Key[:0], []byte("host")...)
+	tag.Value = append(tag.Value[:0], v...)
+
+	v = host.Get("name").GetStringBytes()
+	if len(v) == 0 {
+		return fmt.Errorf("missing `name` element in `host` object")
+	}
+	tag = r.addTag()
+	tag.Key = append(tag.Key[:0], []byte("hostname")...)
+	tag.Value = append(tag.Value[:0], v...)
+
+	v = o.GetStringBytes("name")
+	if len(v) == 0 {
+		return fmt.Errorf("missing `name` element")
+	}
+	tag = r.addTag()
+	tag.Key = append(tag.Key[:0], []byte("__name__")...)
+	tag.Value = append(tag.Value[:0], v...)
+
+	n, err := getFloat64(o, "value")
+	if err != nil {
+		return fmt.Errorf("missing `value` element, %s", err)
+	}
+	r.Value = n
+
+	cl, err := getInt64(o, "clock")
+	if err != nil {
+		return fmt.Errorf("missing `clock` element, %s", err)
+	}
+	ns, err := getInt64(o, "ns")
+	if err != nil {
+		return fmt.Errorf("missing `ns` element, %s", err)
+	}
+	// clock - Number of seconds since Epoch to the moment when value was collected (integer part).
+	// ns - Number of nanoseconds to be added to clock to get a precise value collection time.
+	//
+	// See https://www.zabbix.com/documentation/current/en/manual/appendix/protocols/real_time_export#item-values
+	r.Timestamp = cl*1e3 + ns/1e6
+
+	if (par & ADDGROUPS) != 0 {
+		groups, err := getArray(o, "groups")
+		if err != nil {
+			return fmt.Errorf("missing `groups` element, %s", err)
+		}
+		for _, g := range groups {
+			k := g.GetStringBytes()
+			if len(k) == 0 {
+				continue
+			}
+
+			tag = r.addTag()
+			tag.Key = append(tag.Key[:0], k...)
+			tag.Value = append(tag.Value[:0], EMPTYVALUE)
+		}
+	}
+
+	skipEmptyTags := (par & ADDEMPTYTAGS) == 0
+	item_tags, err := getArray(o, "item_tags")
+	if err != nil {
+		return fmt.Errorf("missing `item_tags` element, %s", err)
+	}
+	for _, t := range item_tags {
+		k := t.GetStringBytes("tag")
+		if len(k) == 0 {
+			continue
+		}
+
+		v := t.GetStringBytes("value")
+		if len(v) == 0 && skipEmptyTags {
+			continue
+		}
+		tag = r.addTag()
+		tag.Key = append(tag.Key[:0], k...)
+		if len(v) == 0 {
+			tag.Value = append(tag.Value[:0], EMPTYVALUE)
```

Comment:
Could you share a use-case when empty labels must be kept?

Reply:
Zabbix allows tags without values. Various filtering rules and conditions in Zabbix also support tags without values. Perhaps someone might use this feature and want to see similar tags as labels in VicotriaMetrics.

I missed something when implementing Zabbix Connector support. Zabbix supports the same tag names with different values for the same element. For example:
```
service
service: mysql 
service: critical
```

How could this be implemented in VicotriaMetrics? (`service: empty,mysql,critical`)

https://www.zabbix.com/documentation/current/en/manual/config/tagging
https://www.zabbix.com/documentation/current/en/manual/config/notifications/action/conditions

Reply:
For such tags I would suggest creating separate labels:
```
tag_service_mysql: "present"
tag_service_critical: "present"

# or 

# this is similar to how k8s discovery handles such cases
tag_present_service_mysql: "true"
tag_present_service_critical: "true"
```

Reply:
After much discussion and testing, groups and tags are now handled like this:

- The prefix `group_` is added to group names if the `zabbixconnector.addGroups` flag is specified.
- Tag names are appended with the prefix `tag_`.
- Added flag `zabbixconnector.mergeDuplicateTags=<value>`. It involves merging similar tags. The values ​​are concatenated with a `<value>` delimiter under the common label name.